prow/gerrit/fakegerrit/fakegerrit.go

```diff
@@ -0,0 +1,131 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package fakegerrit
+
+import (
+	"sync"
+
+	gerrit "github.com/andygrunwald/go-gerrit"
+)
+
+type Project struct {
+	Branches  map[string]*gerrit.BranchInfo
+	ChangeIDs []string
+}
+
+type Change struct {
+	ChangeInfo *gerrit.ChangeInfo
+	Comments   map[string][]*gerrit.CommentInfo
+}
+
+type FakeGerrit struct {
+	Changes  map[string]Change
+	Accounts map[string]*gerrit.AccountInfo
+	Projects map[string]*Project
+	// lock to be thread safe
+	lock sync.Mutex
+}
+
+func (fg *FakeGerrit) Reset() {
+	fg.lock.Lock()
+	defer fg.lock.Unlock()
+
+	fg.Changes = make(map[string]Change)
+	fg.Accounts = make(map[string]*gerrit.AccountInfo)
+	fg.Projects = make(map[string]*Project)
+}
+
+func (fg *FakeGerrit) GetChangesForProject(projectName string, start int) []*gerrit.ChangeInfo {
+	res := []*gerrit.ChangeInfo{}
+	for _, id := range fg.Projects[projectName].ChangeIDs {
+		if start > 0 {
```

Comment:
Can you comment on the desired behavior? E.g., if we call GetChangesForProject("foo", 20), then we expect to get all changes 20, 21, 22, 23, etc ?

If that's the case then it's probably better to pass in a 3rd variable, `end`, to define some range between `start` and `end` (and maybe let `end` be `nil` if we want to get from `start` to the end of the range).

Reply:
Also, do we need to `fg.lock.Lock()` and unlock here like in `GetComments`?

Reply:
Added a comment, and a total (which will be more useful then end as the api can take a number of changes it will want to return).

Reply:
Added a comment, and a total (which will be more useful then end as the api can take a number of changes it will want to return).

Reply:
The lock is in getChange. This is just a helper that calls it for each change in the project.

Reply:
The lock is in getChange. This is just a helper that calls it for each change in the project.

Reply:
`Starts at project number start` should probably be reworded to:

```
Skips the first `start` number of ChangeIDs. `total` caps the total to a number smaller or equal to the actual total number of ChangeIDs.
```

Am I understanding this correctly? If so, I think `total` is not a good name because there are 2 totals --- the desired total (provided as the argument here) and the actual total (the range we're iterating through in `fg.Projects[projectName].ChangeIDs`). Maybe `desired_total` is better?

It's also not clear why we want to `start--` to skip `start` number of changes (maybe that should be in the comments also?). Since `start` is used to skip past an offset, `offset` or `skip` is probably a better name.

Reply:
Done. Leaving variable name as start though as that is what it is called in the Gerrit API